pkg/cache/cache.go

```diff
@@ -0,0 +1,158 @@
+package cache
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+var DefaultCacheExpiration = time.Now().Add(30 * time.Minute)
+
+const (
+	RosaConfigDir = "ROSA_CONFIG_DIR"
+	GobName       = "ocm-cache.gob"
+)
+
+//go:generate mockgen -source=cache.go -package=cache -destination=./cache_mock.go
+type RosaCache interface {
+	Set(k string, x interface{}, d time.Time)
+	Get(k string) (interface{}, bool)
+	Items() map[string]Item
+	Dir() (string, error)
+}
+
+var _ RosaCache = &rosaCache{}
+
+type RosaCacheSpec struct {
+	DefaultExpiration time.Time
+}
+
+type rosaCache struct {
+	defaultExpiration time.Time
+	items             map[string]Item
+	mu                sync.RWMutex
+}
+
+func NewRosaCache(spec RosaCacheSpec) RosaCache {
+	cache := &rosaCache{
+		items: make(map[string]Item),
+	}
+	if spec.DefaultExpiration.IsZero() {
+		cache.defaultExpiration = DefaultCacheExpiration
+		return cache
+	}
+	cache.defaultExpiration = spec.DefaultExpiration
+	return cache
+}
+
+func (c *rosaCache) Set(k string, x interface{}, d time.Time) {
+	if d.IsZero() {
+		d = c.defaultExpiration
+	}
+	c.mu.Lock()
+	c.items[k] = Item{
+		Object:     x,
+		Expiration: d,
+	}
+	c.mu.Unlock()
+}
+
+func (c *rosaCache) Get(k string) (interface{}, bool) {
+	c.mu.RLock()
+	item, found := c.items[k]
+	if !found {
+		c.mu.RUnlock()
+		return nil, false
+	}
+	if !item.Expiration.IsZero() {
+		if item.Expired() {
+			c.mu.RUnlock()
+			return nil, false
+		}
+	}
+	c.mu.RUnlock()
+	return item.Object, true
+}
+
+func (c *rosaCache) Items() map[string]Item {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	m := make(map[string]Item, len(c.items))
+	for k, v := range c.items {
+		if !v.Expiration.IsZero() {
+			if v.Expired() {
+				continue
+			}
+		}
+		m[k] = v
+	}
+	return m
+}
+
+func (c *rosaCache) Dir() (string, error) {
+	configDir, hasEnvVar, err := getConfigDirectoryEnvVar()
+	if err != nil {
+		return "", fmt.Errorf("error getting config directory: %v", err)
+	}
+	if hasEnvVar {
+		return configDir, nil
+	}
+
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", fmt.Errorf("error getting current user home dir: %v", err)
+	}
+
+	ocmConfigPath := filepath.Join(homeDir, ".config", "ocm", GobName)
+	ocmConfigPathExists, err := pathExists(ocmConfigPath)
+	if err != nil {
+		return "", fmt.Errorf("error checking if path exists: %v", err)
+	}
+	if ocmConfigPathExists {
+		return ocmConfigPath, nil
+	}
+
+	tmpDir, err := getAndEnsureOCMDirectoryExists()
+	if err != nil {
+		return "", fmt.Errorf("error ensuring tmp OCM directory exists: %v", err)
```

Comment:
@ciaranRoche We seem to mix ROSA and OCM interchangeably. Which one is the most appropriate? I think I'd find it a bit confusing to get an error message about an `OCM` directory when using a tool called ROSA. WDYT?

Reply:
Agree, we do have a mixture, Will update to ROSA